action.yml

```diff
@@ -26,6 +26,196 @@ inputs:
+        DOCKER_IMAGE_NAME=$DOCKER_IMAGE_URL
+        if [[ "$MUST_BUILD" == "true" ]]; then
+          DOCKER_IMAGE_NAME="${DOCKER_CONTAINER_LABEL}:$(uuidgen)"
+        fi
+
+        # actions/runner source: https://github.com/actions/runner/blob/0484afeec71b612022e35ba80e5fe98a99cd0be8/src/Runner.Worker/Container/DockerCommandManager.cs#L130-L140
+        # Gather list of environment variable names starting with GITHUB_
+        GITHUB_ENV_VARS=$(env | while IFS= read -r line; do key=${line%%=*}; if [[ "$key" =~ ^GITHUB_.+$ ]]; then printf -- '-e %s ' "$key"; fi; done)
+        # Gather list of environment variable names starting with ACTIONS_
+        ACTIONS_ENV_VARS=$(env | while IFS= read -r line; do key=${line%%=*}; if [[ "$key" =~ ^ACTIONS_.+$ ]]; then printf -- '-e %s ' "$key"; fi; done)
+        # Gather list of environment variable names with values starting with INPUT_
+        INPUT_ENV_VARS=$(env | while IFS= read -r line; do key=${line%%=*}; val=${line#*=}; if [[ "$key" =~ ^INPUT_.+$ ]]; then printf -- '-e %s=%s ' "$key" "$val"; fi; done)
+        PASSTHROUGH_ENV_VARS="${GITHUB_ENV_VARS}${ACTIONS_ENV_VARS}${INPUT_ENV_VARS}"
+
+        # Set step outputs
+        echo "building=${MUST_BUILD}" >> $GITHUB_OUTPUT
+        echo "image_url=${DOCKER_IMAGE_URL}" >> $GITHUB_OUTPUT
+        echo "image_name=${DOCKER_IMAGE_NAME}" >> $GITHUB_OUTPUT
+        echo "container_name=${DOCKER_CONTAINER_NAME}" >> $GITHUB_OUTPUT
+        echo "container_label=${DOCKER_CONTAINER_LABEL}" >> $GITHUB_OUTPUT
+        echo "env_expansion=${PASSTHROUGH_ENV_VARS}" >> $GITHUB_OUTPUT
+
+    - if: ${{ steps.docker.outputs.building != 'true' }}
+      name: Pull prebuilt Docker image
+      shell: sh
+      env:
+        DOCKER_IMAGE_URL: ${{ steps.docker.outputs.image_url }}
+      run: |
+        docker pull $DOCKER_IMAGE_URL
+
+    - if: ${{ steps.docker.outputs.building == 'true' }}
+      name: Build local Docker image
+      shell: sh
+      env:
+        DOCKER_IMAGE_NAME: ${{ steps.docker.outputs.image_name }}
+      run: |
+        # Manually change directory to avoid bug with github.action_path reference
+        # Issue: https://github.com/actions/runner/issues/716
+        # Pending PR: https://github.com/actions/runner/pull/2517
+        cd $GITHUB_ACTION_PATH
+
+        # NOTE: This will rely on DockerHub to pull the ruby-slim image
+        docker build -t $DOCKER_IMAGE_NAME -f Dockerfile .
+
+    - name: Run task in Docker container
+      shell: sh
+      env:
+        DOCKER_CONTAINER_NAME: ${{ steps.docker.outputs.container_name }}
+        DOCKER_CONTAINER_LABEL: ${{ steps.docker.outputs.container_label }}
+        DOCKER_IMAGE_NAME: ${{ steps.docker.outputs.image_name }}
+      run: |
+        docker run \
+          --name $DOCKER_CONTAINER_NAME \
+          --label $DOCKER_CONTAINER_LABEL \
+          --workdir /github/workspace \
+          --rm \
+          ${{ steps.docker.outputs.env_expansion }} \
```

Comment:
This has to be plopped directly into the string interpolation/evaluation here instead of being passed in as an environment variable in order to allow the `docker run` command to treat it as multiple individual arguments.